traits/observers/_has_traits_helpers.py

```diff
@@ -27,6 +31,53 @@
 ]
 
 
+def warn_comparison_mode(object, name):
+    """ Check if the trait is a Dict/List/Set with comparison_mode set to
+    equality (or higher). If so, warn about the fact that observers will be
+    lost in the event of reassignment.
+
+    Parameters
+    ----------
+    object : HasTraits
+        Object where a trait is defined
+    name : str
+        Name of a trait to check if warning needs to be issued for it.
+    """
+    ctrait = object.traits()[name]
+
+    def has_container_trait(ctrait):
+        # Return true if the CTrait trait type contains container trait
+        # at the top-level.
+        container_types = (Dict, List, Set)
+        is_container = ctrait.is_trait_type(container_types)
+        if is_container:
+            return True
+        # Try inner traits, e.g. to support Union
+        return any(
+            trait.is_trait_type(container_types)
+            for trait in ctrait.inner_traits
+        )
+
+    if (not ctrait.is_property
+            and has_container_trait(ctrait)
+            and ctrait.comparison_mode > ComparisonMode.identity):
```

Comment:
We shouldn't rely on ordering being meaningful for `ComparisonMode` entries. Let's instead check for `ComparisonMode.equality` directly.  If you want to make this code more defensive against possible future additions to `ComparisonMode`, you could raise `NotImplementedError` for comparison modes other than the known three.